File: solutions/typescript/scrabble-score/1/scrabble-score.ts
export function score(word: string | undefined): number {
  // 1. Manejo de casos nulos o vacíos
  if (!word) {
    return 0;
  }

  // 2. Definición del diccionario de valores
  // Usamos Record<string, number> para tipar el objeto en TypeScript
  const letterValues: Record<string, number> = {
    'A': 1, 'E': 1, 'I': 1, 'O': 1, 'U': 1, 'L': 1, 'N': 1, 'R': 1, 'S': 1, 'T': 1,
    'D': 2, 'G': 2,
    'B': 3, 'C': 3, 'M': 3, 'P': 3,
    'F': 4, 'H': 4, 'V': 4, 'W': 4, 'Y': 4,
    'K': 5,
    'J': 8, 'X': 8,
    'Q': 10, 'Z': 10
  };

  // 3. Cálculo de la puntuación
  // Convertimos a mayúsculas, separamos en letras y sumamos
  return word
    .toUpperCase()
    .split('')
    .reduce((total, letter) => {
      // Si la letra existe en el mapa, sumamos su valor, si no (como un espacio), sumamos 0
      return total + (letterValues[letter] || 0);
    }, 0);
}

File: solutions/ruby/scrabble-score/1/scrabble_score.rb
class Scrabble
  # 1. Definimos la tabla de valores de forma agrupada
  VALUES = {
    %w[A E I O U L N R S T] => 1,
    %w[D G]                => 2,
    %w[B C M P]            => 3,
    %w[F H V W Y]          => 4,
    %w[K]                  => 5,
    %w[J X]                => 8,
    %w[Q Z]                => 10
  }

  # 2. Transformamos esa tabla en un diccionario simple: { 'A' => 1, 'B' => 3 ... }
  # Usamos each_with_object para construir el nuevo Hash
  LETTER_VALUES = VALUES.each_with_object({}) do |(letters, value), hash|
    letters.each { |letter| hash[letter] = value }
  end

  def initialize(word)
    # Guardamos la palabra en mayúsculas y limpiamos espacios si los hay
    @word = word.to_s.upcase.strip
  end

  def score
    # 3. Convertimos la palabra en letras, buscamos su valor y sumamos
    # Si la letra no existe en nuestro diccionario, el valor será 0
    @word.chars.sum { |letter| LETTER_VALUES[letter] || 0 }
  end

  # Los tests de Exercism a veces piden un método de clase directo
  def self.score(word)
    new(word).score
  end
end